.dotnet/src/Custom/Batch/BatchOperation.Protocol.cs

```diff
@@ -0,0 +1,287 @@
+
+        using JsonDocument doc = JsonDocument.Parse(response.Content);
+        string status = doc.RootElement.GetProperty("status"u8).GetString()!;
+
+        return new BatchOperation(client.Pipeline, client.Endpoint, token.BatchId, status, response);
+    }
+
+    /// <inheritdoc/>
+    public override async Task WaitForCompletionAsync(CancellationToken cancellationToken = default)
+    {
+        _pollingInterval ??= new();
+
+        while (!IsCompleted)
+        {
+            PipelineResponse response = GetRawResponse();
+
+            await _pollingInterval.WaitAsync(response, cancellationToken);
+
+            ClientResult result = await GetBatchAsync(_batchId, cancellationToken.ToRequestOptions()).ConfigureAwait(false);
+
+            ApplyUpdate(result);
+        }
+    }
+
+    /// <summary>
+    /// Waits for the operation to complete processing on the service.
+    /// </summary>
+    /// <param name="pollingInterval"> The time to wait between sending requests
+    /// for status updates from the service. </param>
+    /// <param name="cancellationToken"> A token that can be used to cancel this
+    /// method call. </param>
+    public async Task WaitForCompletionAsync(TimeSpan pollingInterval, CancellationToken cancellationToken = default)
+    {
+        _pollingInterval = new(pollingInterval);
+
+        await WaitForCompletionAsync(cancellationToken).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc/>
+    public override void WaitForCompletion(CancellationToken cancellationToken = default)
+    {
+        _pollingInterval ??= new();
+
+        while (!IsCompleted)
+        {
+            PipelineResponse response = GetRawResponse();
+
+            _pollingInterval.Wait(response, cancellationToken);
+
+            ClientResult result = GetBatch(_batchId, cancellationToken.ToRequestOptions());
+
+            ApplyUpdate(result);
+        }
+    }
+
+    /// <summary>
+    /// Waits for the operation to complete processing on the service.
+    /// </summary>
+    /// <param name="pollingInterval"> The time to wait between sending requests
+    /// for status updates from the service. </param>
+    /// <param name="cancellationToken"> A token that can be used to cancel this
+    /// method call. </param>
+    public void WaitForCompletion(TimeSpan pollingInterval, CancellationToken cancellationToken = default)
+    {
+        _pollingInterval = new(pollingInterval);
+
+        WaitForCompletion(cancellationToken);
+    }
+
+    private void ApplyUpdate(ClientResult result)
+    {
+        PipelineResponse response = result.GetRawResponse();
+
+        using JsonDocument doc = JsonDocument.Parse(response.Content);
+        string? status = doc.RootElement.GetProperty("status"u8).GetString();
+
+        IsCompleted = GetIsCompleted(status);
+        SetRawResponse(response);
+    }
+
+    private static bool GetIsCompleted(string? status)
+    {
+        return status == InternalBatchStatus.Completed.ToString() ||
```

Comment:
I could not find InternalBatchStatus type, but I am assuming it's an extensible enum? If yes, maybe we should just add an == operator to it  so that this line can be return "status == InternalBatchStatus.Completed". THis will look less scary. As of right now, it's not clear whether the code (ToString) allocates or not.

Reply:
Ah, it already exists -- I will use that instead.  Thanks!